### PR TITLE
[Mosaic GPU] Support collective tcgen05.commit arrivals in multidimensional clusters

### LINKED ISSUE #40165: [Mosaic GPU] Support collective tcgen05.commit arrivals in multidimensional clusters
`tcgen05.commit_arrive` currently only accepts a `(2, 1, 1)` CTA cluster for collective commit arrivals due to this line
```python
if ctx.cluster_size != (2, 1, 1):
      raise NotImplementedError("Collective arrivals only support (2, 1, 1)-shaped clusters")
```

Any plans to support multidim clusters such as (2,2,1)? Would a pr for supporting this be welcome? I'd like to explore further optimizations for B200 GEMMs on mosaic kernels and I can reproduce it locally on a B200.

### PR DESCRIPTION
**Pull request by saladpalad**
This pr allows multidim cluster shapes for tcgen05 commit arrival instructions (fixes https://github.com/jax-ml/jax/issues/40165)

- The multicast mask is derived from issuing CTA's linear cluster index `(0b11 << (idx & ~1))`, matching the PTX ISA's CTA-pair definition ([9.7.17.5.1. CTA Pair](https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#tcgen05-cta-pair)).

### CODE REVIEW

**Comment by saladpalad** [comment]
@bchetioui Requesting review. Thanks!

File: jax/experimental/mosaic/gpu/tcgen05.py [hunk 1]
```diff
@@ -854,11 +854,15 @@ def commit_arrive(
   if collective:
     if ctx is None:
       raise ValueError("ctx must be provided for collective barriers")
-    # TODO(apaszke): This is just 0b11 shifted by the even CTA index.
-    if ctx.cluster_size != (2, 1, 1):
-      raise NotImplementedError("Collective arrivals only support (2, 1, 1)-shaped clusters")
+    if ctx.cluster_size[0] % 2:
+      raise ValueError(
+          "Collective arrivals require an even cluster size along the x"
```
Comment (bchetioui) [inline jax/experimental/mosaic/gpu/tcgen05.py]: nit: can you say "minormost cluster dimension" instead of "x" here please?

This may surface in Pallas user code, where the `collective_axis` may have another name.

File: jax/experimental/mosaic/gpu/tcgen05.py [hunk 2]
```diff
@@ -854,11 +854,15 @@ def commit_arrive(
   if collective:
     if ctx is None:
       raise ValueError("ctx must be provided for collective barriers")
-    # TODO(apaszke): This is just 0b11 shifted by the even CTA index.
-    if ctx.cluster_size != (2, 1, 1):
-      raise NotImplementedError("Collective arrivals only support (2, 1, 1)-shaped clusters")
+    if ctx.cluster_size[0] % 2:
```
Comment (bchetioui) [inline jax/experimental/mosaic/gpu/tcgen05.py]: I don't think we should be this loose; the Pallas semantics are such that we expect the minormost axis to always have size `2` for us to be able to issue `tcgen05.commit_arrive`.

Let's restrict this to be `== 2` along the minormost dimension.

File: tests/mosaic/gpu_test.py [hunk 3]
```diff
@@ -3519,6 +3519,106 @@ def quantize(x):
     atol = 2e-2 if out_jax_dtype == jnp.float16 else 5e-6
     np.testing.assert_allclose(z, ref, atol=atol)
 
+  @parameterized.product(cluster=((2, 2, 1), (4, 1, 1), (2, 1, 2)))
+  def test_mma_collective_multidim_cluster(self, cluster):
```
Comment (bchetioui) [inline tests/mosaic/gpu_test.py:3523]: Could we also add tests at the Pallas-level (you can add it in `mosaic_gpu_test.py`, in the sm100-specific tests (`PallasCallTCGen05Test`))?

If it feels redundant, also OK to only test there---the Pallas-level tests seem more important.

**Review by bchetioui — CHANGES_REQUESTED** [review]
Thank you! A few comments.

File: jax/experimental/mosaic/gpu/tcgen05.py [hunk 4]
```diff
@@ -854,11 +854,15 @@ def commit_arrive(
...
+    if ctx.cluster_size[0] != 2:
+      raise ValueError(
+          "Collective arrivals require the minormost cluster dimension to"
+          f" have size 2, got: {ctx.cluster_size}"
+      )
     i16 = ir.IntegerType.get_signless(16)
-    mask = arith.constant(i16, 3)
+    block_idx = arith.index_castui(i16, utils.cluster_idx())
+    even_block_idx = arith.andi(block_idx, arith.constant(i16, ~1))
+    mask = arith.shli(arith.constant(i16, 0b11), even_block_idx)
```
Comment (bchetioui) [inline jax/experimental/mosaic/gpu/tcgen05.py]: Thanks! Just a last nit: this will cost us more ALU operations in the common case, which could be slightly detrimental.

Could we add a fast runtime path

```python
if ctx.cluster_size == (2, 1, 1):
  mask = arith.constant(i16, 3)
```

to make sure we avoid needless computations at runtime in that case?

File: tests/pallas/mosaic_gpu_test.py [hunk 5]
```diff
@@ -7209,6 +7207,77 @@ def kernel(a_gmem, b_gmem, out_gmem, a_smem, b_smem,
...
+          a_smem,
+          b_smem,
+          mma_barrier,
+          accumulate=False,
+          collective_axis="x",
+      )
+      plgpu.barrier_wait(mma_barrier)
+      out_gmem[pair_idx, slice_lhs, :] = plgpu.async_load_tmem(
+          acc_tmem, layout=plgpu.Layout.TCGEN05_M64_COLLECTIVE(n)
+      ).astype(dtype)
```
Comment (bchetioui) [inline tests/pallas/mosaic_gpu_test.py]: nit: is the downcast necessary? You are comparing against a float32 reference.

**Review by bchetioui — APPROVED** [review]
Looks great! A couple of minor comments still, but thank you very much!

File: jax/experimental/mosaic/gpu/tcgen05.py (hunk 4 above)
Comment (saladpalad) [reply jax/experimental/mosaic/gpu/tcgen05.py]: Done. Fwiw testing on B200 w/ `jax-cuda13-pjrt=0.11.2.dev20260831`, when cluster size is (2,1,1), the extra instructions folded into a constant and the ptx still generated `mov.b16 %rs1, 3` w/o the fast runtime path.

File: tests/pallas/mosaic_gpu_test.py (hunk 5 above)
Comment (saladpalad) [reply tests/pallas/mosaic_gpu_test.py]: No its not. Fixed.

**Review by bchetioui — APPROVED** [review]

**Comment by bchetioui** [comment]
Thank you for the change!
